File: haskell/src/Surd/Radical/Denest.hs
-- |
-- Module      : Surd.Radical.Denest
-- Description : Top-level radical denesting dispatcher
-- Stability   : experimental
--
-- Provides the main entry point for radical denesting. The dispatcher
-- applies the appropriate algorithm based on the root index and
-- expression structure:
--
-- 1. For square roots (@n = 2@): tries "Surd.Radical.Denest.Sqrt"
--    (Borodin's algorithm for @sqrt(a + b*sqrt r)@) first, then falls
--    back to Landau's algorithm.
-- 2. For higher roots: tries "Surd.Radical.Denest.NthRoot" (cube root
--    denesting, perfect power extraction, root composition) first,
--    then falls back to Landau's algorithm.
-- 3. Landau's algorithm ("Surd.Radical.Denest.Landau") is the most
--    general fallback: it factors @x^n - a@ over the radical extension
--    field using Trager's algorithm and extracts lower-degree roots.
--
-- All algorithms are applied recursively, bottom-up through the
-- expression tree.
module Surd.Radical.Denest
  ( denest,
    denestFull,
  )
where

import Surd.Radical.Denest.Landau (denestRadical)
import Surd.Radical.Denest.NthRoot (denestNthRoot)
import Surd.Radical.Denest.Sqrt (denestSqrt)
import Surd.Radical.Eval (evalExact)
import Surd.Radical.Normalize (normalize)
import Surd.Types

-- | Denest a radical expression: normalize first, then apply all
-- available denesting algorithms recursively.
--
-- This is the primary entry point for denesting. It normalizes the
-- expression first (to canonicalize like terms, extract perfect powers,
-- etc.) and then runs the full denesting pass.
denest :: RadExpr Rational -> RadExpr Rational
denest = denestFull . normalize

-- | Full denesting pass without pre-normalization.
--
-- Applies specialized algorithms first (Borodin for square roots,
-- cube root denesting for cubic roots), then falls back to Landau's
-- general algorithm. Results are verified via exact real evaluation
-- to ensure correctness.
denestFull :: RadExpr Rational -> RadExpr Rational
denestFull expr = case expr of
  Root 2 inner ->
    let specialized = denestSqrt expr
     in if changed specialized expr
          then specialized
          else tryLandau 2 inner expr
  Root n inner ->
    let specialized = denestNthRoot expr
     in if changed specialized expr
          then specialized
          else tryLandau n inner expr
  Neg a -> Neg (denestFull a)
  Add a b -> Add (denestFull a) (denestFull b)
  Mul a b -> Mul (denestFull a) (denestFull b)
  Inv a -> Inv (denestFull a)
  Pow a n -> Pow (denestFull a) n
  e -> e
  where
    -- Try Landau denesting as a fallback when specialized algorithms don't help
    tryLandau n inner original =
      let inner' = denestFull inner
       in case denestRadical n inner' of
            Just denested ->
              let dv = evalExact denested
                  ov = evalExact original
               in if abs (dv - ov) < 1e-40
                    then denestFull denested -- recurse on result
                    else original
            Nothing -> case original of
              Root _ _ -> Root n (denestFull inner)
              _ -> original
    -- Check if denesting actually changed anything
    changed new old = show new /= show old

File: mercury/denest.m
%---------------------------------------------------------------------------%
% vim: ft=mercury ts=4 sw=4 et
%---------------------------------------------------------------------------%
% File: denest.m
% Main author: nyc
%
% Top-level radical denesting dispatcher.
%
% Applies the appropriate denesting algorithm based on the root index
% and expression structure:
% - Square root denesting (Borodin-FHKT)
% - General nth-root simplification and cube root denesting
% - (Landau denesting: planned, not yet implemented)
%
%---------------------------------------------------------------------------%

:- module denest.
:- interface.

:- import_module rad_expr.
:- import_module rational.

%---------------------------------------------------------------------------%

    % Denest a radical expression: normalize first, then apply all
    % available denesting algorithms recursively.
    %
:- func denest(rad_expr(rational)) = rad_expr(rational).

    % Full denesting pass without pre-normalization.
    %
:- func denest_full(rad_expr(rational)) = rad_expr(rational).

%---------------------------------------------------------------------------%
%---------------------------------------------------------------------------%

:- implementation.

:- import_module denest_sqrt.
:- import_module denest_nthroot.
:- import_module rad_normalize.
:- use_module string.

%---------------------------------------------------------------------------%

denest(Expr) = denest_full(normalize(Expr)).

denest_full(Expr) = Result :-
    ( if Expr = re_root(2, Inner) then
        Specialized = denest_sqrt.denest_sqrt(Expr),
        ( if changed(Specialized, Expr) then
            Result = Specialized
        else
            Result = re_root(2, denest_full(Inner))
        )
    else if Expr = re_root(N, Inner) then
        Specialized = denest_nthroot.denest_nthroot(Expr),
        ( if changed(Specialized, Expr) then
            Result = Specialized
        else
            Result = re_root(N, denest_full(Inner))
        )
    else if Expr = re_neg(A) then
        Result = re_neg(denest_full(A))
    else if Expr = re_add(A, B) then
        Result = re_add(denest_full(A), denest_full(B))
    else if Expr = re_mul(A, B) then
        Result = re_mul(denest_full(A), denest_full(B))
    else if Expr = re_inv(A) then
        Result = re_inv(denest_full(A))
    else if Expr = re_pow(A, N) then
        Result = re_pow(denest_full(A), N)
    else
        Result = Expr
    ).

:- pred changed(rad_expr(rational)::in, rad_expr(rational)::in) is semidet.

changed(New, Old) :-
    New \= Old.

%---------------------------------------------------------------------------%
:- end_module denest.
%---------------------------------------------------------------------------%
